crypto/ec/ec2_smpl.c

```diff
@@ -737,4 +680,275 @@ int ec_GF2m_simple_field_div(const EC_GROUP *group, BIGNUM *r,
+                               EC_POINT *p, BN_CTX *ctx)
+{
+    int ret = 0;
+    BIGNUM *t0, *t1, *t2 = NULL;
+
+    if (BN_is_zero(r->Z))
+        return EC_POINT_set_to_infinity(group, r);
+
+    if (BN_is_zero(s->Z)) {
+        if (!EC_POINT_copy(r, p)
+            || !EC_POINT_invert(group, r, ctx)) {
+            ECerr(EC_F_EC_GF2M_SIMPLE_LADDER_POST, ERR_R_EC_LIB);
+            return 0;
+        }
+        return 1;
+    }
+
+    BN_CTX_start(ctx);
+    t0 = BN_CTX_get(ctx);
+    t1 = BN_CTX_get(ctx);
+    t2 = BN_CTX_get(ctx);
+    if (t2 == NULL) {
+        ECerr(EC_F_EC_GF2M_SIMPLE_LADDER_POST, ERR_R_MALLOC_FAILURE);
+        goto err;
+    }
+
+    if (!group->meth->field_mul(group, t0, r->Z, s->Z, ctx)
+        || !group->meth->field_mul(group, t1, p->X, r->Z, ctx)
+        || !BN_GF2m_add(t1, r->X, t1)
+        || !group->meth->field_mul(group, t2, p->X, s->Z, ctx)
+        || !group->meth->field_mul(group, r->Z, r->X, t2, ctx)
+        || !BN_GF2m_add(t2, t2, s->X)
+        || !group->meth->field_mul(group, t1, t1, t2, ctx)
+        || !group->meth->field_sqr(group, t2, p->X, ctx)
+        || !BN_GF2m_add(t2, p->Y, t2)
+        || !group->meth->field_mul(group, t2, t2, t0, ctx)
+        || !BN_GF2m_add(t1, t2, t1)
+        || !group->meth->field_mul(group, t2, p->X, t0, ctx)
+        || !BN_GF2m_mod_inv(t2, t2, group->field, ctx)
+        || !group->meth->field_mul(group, t1, t1, t2, ctx)
+        || !group->meth->field_mul(group, r->X, r->Z, t2, ctx)
+        || !BN_GF2m_add(t2, p->X, r->X)
+        || !group->meth->field_mul(group, t2, t2, t1, ctx)
+        || !BN_GF2m_add(r->Y, p->Y, t2)
+        || !BN_one(r->Z))
+        goto err;
+
+    r->Z_is_one = 1;
+
+    /* GF(2^m) field elements should always have BIGNUM::neg = 0 */
+    BN_set_negative(r->X, 0);
+    BN_set_negative(r->Y, 0);
+
+    ret = 1;
+
+ err:
+    BN_CTX_end(ctx);
+    return ret;
+}
+
+static
+int ec_GF2m_simple_points_mul(const EC_GROUP *group, EC_POINT *r,
+                              const BIGNUM *scalar, size_t num,
+                              const EC_POINT *points[],
+                              const BIGNUM *scalars[],
+                              BN_CTX *ctx)
+{
+    int ret = 0;
+    EC_POINT *t = NULL;
+
+    /*-
+     * We limit use of the ladder only to the following cases:
+     * - r := scalar * G
+     *   Fixed point mul: scalar != NULL && num == 0;
+     * - r := scalars[0] * points[0]
+     *   Variable point mul: scalar == NULL && num == 1;
+     * - r := scalar * G + scalars[0] * points[0]
+     *   used, e.g., in ECDSA verification: scalar != NULL && num == 1
+     *
+     * In any other case (num > 1) we use the default wNAF implementation.
+     *
+     * We also let the default implementation handle degenerate cases like group
+     * order or cofactor set to 0.
+     */
+    if (num > 1 || BN_is_zero(group->order) || BN_is_zero(group->cofactor))
+        return ec_wNAF_mul(group, r, scalar, num, points, scalars, ctx);
```

Comment:
If I understand correctly, you are suggesting to handle the (`scalar != NULL && num==1`) in `ec_wNAF_mul()` calling `ec_scalar_mul_ladder`: the problem with that is that for other implementations `ec_scalar_mul_ladder` can be slower than the regular "non-const-time" path.

Double multiplication is basically only used by ECDSA verification right now, where all the inputs are public, so a fast and insecure implementation could be used, rather than `ec_scalar_mul_ladder`.

What we really would need is refactor the `EC_METHOD` API to ditch/deprecate the unified `EC_POINTs_mul` (which makes a sense from a mathematical point of view, but is not really the way the 3 different scalar multiplication cases are used by cryptosystems) in favor of 3 different function pointers for `fixed_point_mul`, `variable_point_mul`, `double_point_mul`.
Right now we have wrappers that are forcing the 3 different operations into the unified EC_POINTs_mul path, and then in `ec_scalr_mul_ladder` we split them again.
A specific `double_point_mul` could almost half the current execution cost of the verification step, as we could use formulas (e.g., derived from Shamir's trick) that avoid doing 2 distinct scalar multiplications and then 1 addition, and share as much computation as possible.
But this must wait for the next major release development cycle!

Does this answer your comment, or did I completely misunderstood what you meant?

Reply:
> the problem with that is that for other implementations ec_scalar_mul_ladder can be slower

Fair enough. Then how about
```
if (scalar == NULL || num != 1 || order == 0 || cofactor == 0)
    return ec_wNAF_mul()
/* scalar != NULL && num == 1 is faster with ec_scalar_mul_ladder here [and now] */
...
```
Just keep in mind that this is side note, so "how about" is really just a question, it's not a suggestion. Point is not larger than order != 0 && cofactor !=0 case in ec_wNAF_mul being partial duplicate of the code below this line. [Off topic, this kind of asks for question if ec_wNAF_mul is actually suitable name.]

Reply:
I think this change would hinder readability and mask intent: what we want to convey is that binary non-degenerate (i.e., have a `order` and `cardinality` that makes sense, given that mathematically speaking 0 is not really an option for either value) curves use `ec_scalar_mul_ladder` for the 3 cases, any other crazyness that can be fed through the `EC_POINTs_mul` API is basically undocumented and not cryptographically relevant, so we will let the default wNAF implementation handle it.

### Re: renaming `ec_wNAF_mul`

I would argue that it would make sense to have in `ec_mult.c` something like `ec_points_mul_default` working as a wrapper distinguishing the two single point multiplication cases (for which `ec_scalar_mul_ladder` is then called) from the truly multi-point multiplication case (for which `ec_wNAF_mul`) is called.
But I would leave this to a separate PR.

Even better, in a separate PR I would rather start a transition to 3 separate function pointers in `EC_METHOD` for the fixed/variable/double point multiplication, and let the `EC_POINTs_mul` wrapper do something like:

```c
/* All kind of checks on not NULL args, every point_is_compat, etc. */

if (group->meth->points_mul)
    return group->meth->points_mul(...);

degenerate = BN_is_zero(group->order) || BN_is_zero(group->cofactor);

if (scalar != NULL && num == 0)
    if (group->meth->fixed_scalar_mul)
        return group->meth->fixed_scalar_mul(...);
    else if (!degenerate)
        return ec_scalar_mul_ladder(...);

if (scalar == NULL && num == 1)
    if (group->meth->variable_scalar_mul)
        return group->meth->variable_scalar_mul(...);
    else if (!degenerate)
        return ec_scalar_mul_ladder(...);

if (scalar != NULL && num == 1 && group->meth->double_scalar_mul)
    return group->meth->double_scalar_mul(...);

/* HIC SUNT DRACONES */
return ec_wNAF_mul(...);
```
Until something similar exist, I don't think we can avoid specific `EC_POINTs_mul` implementations to repeat the checks for the 3 cases.
And the existence of the degenerate case will also probably means that most non-default implementations of the 3 scalar_mul variants will need to repeat the check for degenerate cases.